src/sysusers/sysusers.c

```diff
@@ -370,341 +370,416 @@ static int rename_and_apply_smack(const char *temp_path, const char *dest_path)
                         goto finish;
                 }
 
-                HASHMAP_FOREACH(i, todo_gids, iterator) {
-                        struct group n = {
-                                .gr_name = i->name,
-                                .gr_gid = i->gid,
-                                .gr_passwd = (char*) "x",
-                        };
+        } else if (errno != ENOENT) {
+                r = -errno;
+                goto finish;
+        } else if (fchmod(fileno(passwd), 0644) < 0) {
+                r = -errno;
+                goto finish;
+        }
 
-                        r = putgrent_with_members(&n, group);
-                        if (r < 0)
-                                goto finish;
+        HASHMAP_FOREACH(i, todo_uids, iterator) {
+                struct passwd n = {
+                        .pw_name = i->name,
+                        .pw_uid = i->uid,
+                        .pw_gid = i->gid,
+                        .pw_gecos = i->description,
 
-                        group_changed = true;
-                }
+                        /* "x" means the password is stored in the shadow file */
+                        .pw_passwd = (char*) "x",
 
-                r = fflush_and_check(group);
-                if (r < 0)
-                        goto finish;
+                        /* We default to the root directory as home */
+                        .pw_dir = i->home ? i->home : (char*) "/",
 
-                if (original) {
-                        fclose(original);
-                        original = NULL;
-                }
+                        /* Initialize the shell to nologin, with one exception:
+                         * for root we patch in something special */
+                        .pw_shell = i->uid == 0 ? (char*) "/bin/sh" : (char*) "/sbin/nologin",
+                };
 
-                /* OK, now also update the shadow file for the group list */
-                gshadow_path = prefix_roota(arg_root, "/etc/gshadow");
-                r = fopen_temporary_label("/etc/gshadow", gshadow_path, &gshadow, &gshadow_tmp);
-                if (r < 0)
+                errno = 0;
+                if (putpwent(&n, passwd) != 0) {
+                        r = errno ? -errno : -EIO;
                         goto finish;
+                }
+        }
 
-                original = fopen(gshadow_path, "re");
-                if (original) {
-                        struct sgrp *sg;
+        r = fflush_and_check(passwd);
 
-                        r = sync_rights(original, gshadow);
-                        if (r < 0)
-                                goto finish;
+finish:
+        if (r < 0) {
+                unlink(passwd_tmp);
+                return r;
+        }
 
-                        errno = 0;
-                        while ((sg = fgetsgent(original))) {
+        *tmpfile = passwd;
+        *tmpfile_path = passwd_tmp;
+        passwd = NULL;
+        passwd_tmp = NULL;
+        return 0;
+}
```

Comment:
this really looks like you want two separate goto targets here: "finish" for the succesfull case, and "fail" or so for the falure case. After all the success and error code paths here share exactly nothing...